### ISSUE TITLE
Confusing error message from spirv-val

### ISSUE THREAD

**Issue by MaikKlein** [KhronosGroup/SPIRV-Tools#1948]
The spirv gist can be found [here](https://gist.github.com/MaikKlein/7a0fb8553eef617dfd431989d1a5f0d4) but I also created a graph from the spv file which I think is easier to read.

![](https://i.imgur.com/c4C3Cgv.png)

I get the following error message: 

> error: line 158: block <ID> 81 branches to the selection construct, but not to the selection header <ID> 84
  %bb1_0 = OpLabel

Is the error message actually correct? I probably violated some rule but I don't think that `%81` is the problem. The actual error seems to originate from `%84`, if I remove the `SelectionMerge` the validator is silent again. But why?

>a selection construct: the set of blocks dominated by a selection header, minus the set of blocks dominated by the header’s merge block

I think that spirv-val thinks that `82` is a selection construct of `84`, which I don't think is correct because `82` dominates `84`. Although I have noticed that `glslang` also sometimes leaves out a selection merge inside a loop, so maybe I am just missing something?

**Comment by alan-baker** [comment]
I agree the error is confusing. I will look into improvements. The real problem appears to be that the OpLoopMerge declares %88 as the merge block so the continue block is an invalid backedge. If you were declare the loop merge as %83 (bb3), the code would validate.

**Comment by alan-baker** [comment]
I agree that the rules in the spec does not invalidate this example as they are written currently.  We are looking into this, but will have to resolve it internally within Khronos first. I will update this issue when a resolution is reached.

In the interim, I would suggest (unofficially) that the continue target of the loop should not be dominated by the loop's merge block.

**Comment by dneto0** [comment]
> Where block 25 doesn't have a SelectionMerge at all.

Yes.  I think it's ok that block 25 does not have a selection merge.  The intuition is that one of the branches out from 25 is a valid exit out of the loop construct to the merge block of the loop.  So even though it's a two-way branch it's not *introducing* a *new* divergence in the control flow, instead it's *resolving* the previous divergence that already occurred at block 21.

However, I personally think the spec could be improved to state the rule clearly.  This is the subject of an ongoing discussion within Khronos.

**Comment by alan-baker** [comment]
> What is very strange is that the merge block from the loop bb2 and from the selection bb4 have to be same or my driver just crashes, but that violates the spec because every merge block should be unique. Maybe it is just a driver bug.

That definitely sounds like a driver bug.
